### PR TITLE
Add alloc-fmt crate for allocation-safe formatting and output

### PR DESCRIPTION
**Pull request by joshlf**
This new `alloc-fmt` crate provides formatting macros equivalent to the standard library's macros, except they don't allocate. I had long suspected that this might be an issue, but I finally came across a bug in PR #85 that is caused by this problem, so I figured it was finally time to address the it.

### CODE REVIEW

File: alloc-fmt/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+README.md file at the top-level directory of this repository.
+
+Licensed under the Apache License, Version 2.0 (the LICENSE-APACHE file) or
+the MIT license (the LICENSE-MIT file) at your option. This file may not be
+copied, modified, or distributed except according to those terms. -->
+
+alloc-fmt
+=========
+
+`alloc-fmt` provides formatting and assertion macros similar to `println`,
```
Comment (ezrosent) [inline alloc-fmt/README.md]: Maybe reword this to indicate that when you say "in an allocator" you mean "global allocator" in particular. I think it's clear from context but we shouldn't give the impression that more specific allocators should worry about println

File: alloc-fmt/src/bin/assert.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+//
+// Licensed under the Apache License, Version 2.0 (the LICENSE-APACHE file) or
+// the MIT license (the LICENSE-MIT file) at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+#[macro_use]
+extern crate alloc_fmt;
+
+#[cfg_attr(feature = "cargo-clippy", allow(cyclomatic_complexity))]
+fn main() {
```
Comment (ezrosent) [inline alloc-fmt/src/bin/assert.rs]: For these binaries, do you think it's worth setting #[global_allocator] to something that panics in alloc, or is that overkill

File: alloc-fmt/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+//!
+//! # Usage and Behavior
+//! The macros in this crate are named `alloc_xxx`, where `xxx` is the name of the equivalent
+//! standard library macro (e.g., `alloc_println`, `alloc_debug_assert`, etc).
+//!
+//! The behavior of the formatting macros is identical to the behavior of their standard library
+//! counterparts. The behavior of the assertion macros is somewhat different. When an assertion
+//! fails, a message is first unconditionally printed to stderr (in case further processing causes
+//! a crash). A stack trace is then printed, and the process aborts. If, during the printing of the
+//! stack trace, recursion is detected (computing stack traces can cause allocation, which can
```
Comment (ezrosent) [inline alloc-fmt/src/lib.rs]: The check for recursion is conservative, no? 

Maybe reword as "If possible recursion is detected while printing the stack trace (e.g. if creating the stack trace results in a call to `alloc` which in turn panics ... ) the process aborts immediately"

File: alloc-fmt/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        unsafe {
+            let guard = $mtx.lock();
+
+            let mut fd = $crate::FDWriter($file);
+            use $crate::Write;
+            let _ = write!(&mut fd, $($arg)*).map_err(|_| {
+                $crate::abort();
+            });
+
+            // explicitly drop the guard so we're guaranteed it lives at least this long;
```
Comment (ezrosent) [inline alloc-fmt/src/lib.rs]: Really? That seems wrong... I thought Mutex<()> was a not-uncommon pattern.

File: alloc-fmt/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+    ($fmt:expr, $($arg:tt)*) => (alloc_eprint!(concat!($fmt, "\n"), $($arg)*));
+}
+
+// Sometimes backtraces allocate. If that allocation causes an assertion failure, then we can get
+// into an infinite recursion scenario. In order to detect this, we set this global bool to true,
+// and load it before printing a backtrace. If we find that it is true, we immediately abort
+// (essentially short-circuiting what would eventually happen if print_backtrace_and_abort
+// successfully finished executing).
+#[doc(hidden)]
+pub static IS_ASSERTING: AtomicBool = ATOMIC_BOOL_INIT;
```
Comment (ezrosent) [inline alloc-fmt/src/lib.rs]: Potential TODO:

have a static array of first N prime numbers, make this an AtomicUsize and have each thread multiply IS_ASSERTING by primes[thread::current().id() % primes.size()] and divide when finished. The check then asks when IS_ASSERTING is divisible by this prime; cuts down on false positives.

File: alloc-fmt/src/bin/assert.rs (hunk 2 above)
Comment (joshlf) [reply alloc-fmt/src/bin/assert.rs]: I think that's overkill; I've tested that functionality by merging this with the current `tls` branch and seeing what happens since that branch has existing assertion failures.

File: alloc-fmt/src/lib.rs (hunk 4 above)
Comment (joshlf) [reply alloc-fmt/src/lib.rs]: I'm not positive, but better safe than sorry.
